### PR TITLE
[lesson2] added async/await and promise solutions for github api task

### PR DESCRIPTION
**Pull request by wizzardio**
Added two files:
 - async.js - solution with async/await;
 - promise.js - solution with promises;

### CODE REVIEW

File: lesson2_async_await/async.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    headers: {
+        'User-Agent': 'Request-Promise'
+    },
+    json: true
+}
+
+const getContributors = async (owner, repo) => {
+    let contributors = [];
+    try {
+        options.uri = util.format(CONTRIBUTORS_URI, owner, repo);
```
Comment (webwin0) [inline lesson2_async_await/async.js]: Modification of a global object is bad practice.
You can use
```
const defaultOptions = {
    headers: {
        'User-Agent': 'Request-Promise'
    },
    json: true
}
```
and then inside function use
```
const options = {...defaultOptions, uri: `${GITHUB_URI}/repos/${owner}/${repo}/contributors`};
```

File: lesson2_async_await/async.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    try {
+        const repos = await getRepos(gitHubOrgName);
+        const results = repos.map(async (repo) => {
+            const contributors = await getContributors(repo.owner, repo.name);
+            const res = {};
+            res[repo.name] = contributors;
+            return res;
+        });
+
+        const resolved = await Promise.all(results); // small hack :)
```
Comment (webwin0) [inline lesson2_async_await/async.js]: It's single solution with async/await to make parallel calls. It's absolutely correct solution.

File: lesson2_async_await/promise.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+const options = {
+    headers: {
+        'User-Agent': 'Request-Promise'
+    },
+    json: true
+}
+
+const getContributors = (owner, repo) => {
+    options.uri = util.format(CONTRIBUTORS_URI, owner, repo);
+    return new Promise((resolve, reject) => {
```
Comment (webwin0) [inline lesson2_async_await/promise.js]: We need wrap with Promise when we work with callbacks and want to promisify interface.
You can return `rp(options)`.

File: lesson2_async_await/promise.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        rp(options)
+            .then((contributors) => {
+                resolve(contributors.map(c => c.login));
```
Comment (webwin0) [inline lesson2_async_await/promise.js]: without Promise wrapper, you need simply `return contributors.map(c => c.login);`

File: lesson2_async_await/promise.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+                return {name: repo.name, owner: repo.owner.login};
+            });
+            resolve(reposData)
+        })
+        .catch(err => reject(err));
+    });
+}
+
+const main = (gitHubOrgName) => {
+    return new Promise((resolve, reject) => {
```
Comment (webwin0) [inline lesson2_async_await/promise.js]: new Promise is not required, return `getRepos`.

File: lesson2_async_await/promise.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        getRepos(gitHubOrgName)
+            .then((repos) => {
+                const promises = repos.map((repo) => {
+                    let promise = new Promise((resolve, reject) => {
```
Comment (webwin0) [inline lesson2_async_await/promise.js]: new Promise is not required, return `getContributors`.

File: lesson2_async_await/promise.js (hunk 3 above)
Comment (wizzardio) [reply lesson2_async_await/promise.js]: Thanks, looks like I confused myself.

File: lesson2_async_await/promise.js (hunk 6 above)
Comment (wizzardio) [reply lesson2_async_await/promise.js]: done

File: lesson2_async_await/async.js (hunk 2 above)
Comment (wizzardio) [reply lesson2_async_await/async.js]: Thanks!

File: lesson2_async_await/promise.js (hunk 4 above)
Comment (wizzardio) [reply lesson2_async_await/promise.js]: Done

File: lesson2_async_await/promise.js (hunk 5 above)
Comment (wizzardio) [reply lesson2_async_await/promise.js]: done

**Comment by wizzardio** [comment]
@webwin0 I fixed code review comments.
Looks like I confused myself with Promise construction. 
Thanks in advance.

**Comment by wizzardio** [comment]
Merged to master

**Pull request closed without merging**